### ISSUE TITLE
Specify order of attribute delete callbacks on MPI_COMM_SELF at MPI_FINALIZE

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#71]
**Originally by _koziol_ on 2008-12-15 15:54:38 -0600**

---

Middleware libraries may have cached information that needs to be flushed/destroyed with MPI calls before the MPI layer goes away (when MPI_Finalize() is called).  [The HDF5 library is an example of this, and needs to flush cached file information to disk]
## History

Revised prior suggestion for an MPI_At_Finalize() routine to use existing callback mechanism on attribute for MPI_COMM_SELF. (3/3/09)

Revised proposal to include line #'s and explicit changes to the advice to implementors section (3/22/09)
## Proposed Solution

These callbacks would be made before any portions of the MPI layer are closed/destroyed, ensuring the callback routines can make any MPI calls necessary.

Change  section 8.7.1 (p283, lines 43-44): from "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in an arbitrary order." to read "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in the reverse order that they were set on MPI_COMM_SELF"

Additionally, we add an advice to implementors note stating the gist of the second part of the "Impact to Implementations" information.  This would modify the text at the top of p.284, lines 1-3 from:

"Advice to implementors. Since attributes can be added from any supported language, the MPI implementation needs to remember the creating language so the correct callback is made. (End of advice to implementors.)"

to read:

"Advice to implementors.  Since attributes can be added from any supported language, the MPI implementation needs to remember the creating language so the correct callback is made.  Implementations that use the attribute delete callback on MPI_COMM_SELF internally should register their internal callbacks before returning from MPI_INIT / MPI_INIT_THREAD, so that libraries or applications will not have portions of the MPI implementation shut down before the application-level callbacks are made.  (End of advice to implementors.)"
## Impact on Implementations

Implementations may need to change how attributes delete callbacks on MPI_COMM_SELF are called, to enforce the LIFO ordering.

Implementations may also need to make certain that any internal implementation use of the attribute delete callback on MPI_COMM_SELF are registered before returning form MPI_INIT/MPI_INIT_THREAD, so that libraries or applications built on top of MPI will not have portions of the MPI implementation shut down before the application-level callbacks are made.
## Impact on Applications / Users

None
## Alternative Solutions

An MPI_At_Finalize() routine (similar to the Standard C atexit() routine) was initially proposed and may still be a valid alternate solution.
## Entry for the Change Log

Section 8.7.1 on page 283.

Attribute deletion callbacks on MPI_COMM_SELF are now called in LIFO order.  Implementors must now also register all implementation-internal attribute deletion callbacks on MPI_COMM_SELF before returning from MPI_INIT/MPI_INIT_THREAD.

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2008-12-16 14:13:10 -0600**

---

Based on feedback from the Forum to build on the existing standard's method of attaching an attribute on MPI_COMM_SELF with a delete callback, I propose dropping the suggested function and instead make the following change to section 8.7.1 (p283):  Change the sentence "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in an arbitrary order." to read "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in the reverse order they were set on MPI_COMM_SELF".

I would also suggest adding a sentence with advise to implementors that reads something like "If an implementation uses this facility to shut down aspects of the implementation, the attributes added by the implementation should be set on MPI_COMM_SELF before returning from MPI_Init, guaranteeing that attributes set on MPI_COMM_SELF by applications will have their delete callback functions executed before the implementation's callback functions."

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2008-12-16 14:17:53 -0600**

---

Add Bill & Jeff as reviewers

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-01-04 18:36:00 -0600**

---

It would be good to also add the rationale for the LIFO ordering of callbacks on COMM_SELF - the rationale is, I believe compelling, and I think it should be captured in the standard.

Bill

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-01-12 20:11:12 -0600**

---

This proposal needs to be converted to use [wiki:mpi22/TicketWorkflow the standard format].

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-03-03 08:54:27 -0600**

---

Updated to conform to standard ticket workflow and to try to capture the rationale for the LIFO ordering.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-03-03 11:24:36 -0600**

---

Just curious -- why is the implementation unnecessary?

Also, in the proposed solution:
- Please list specific line numbers
- If you want to add an ATI, please also include the specific text/page/line numbers/etc. for it.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-03-03 12:14:13 -0600**

---

Agreed on the ATI - the proposed solution needs to be complete and specific.

We do need some implementation - I'll see if MPICH2 already activates the delete callbacks in this order.

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-03-22 19:37:29 -0500**

---

Revised proposal to include line #'s and explicit changes to the advice to implementors section.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2009-03-30 21:34:15 -0500**

---

> Change  section 8.7.1 (p283, lines 43-44): from "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in an arbitrary order." to read "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in the reverse order they were set on MPI_COMM_SELF"

Can you not take care of the order by adding a single callback function for HDF and taking care of everything via extra_state?

> "Advice to implementors.  Since attributes can be added from any supported language, the MPI implementation needs to remember the creating language so the correct callback is made.  Implementations that make internal use of the attribute delete callback on MPI_COMM_SELF should register their internal callbacks before returning from MPI_Init/MPI_Init_Thread(), so that libraries or applications using the implementation will not have portions of the MPI implementation shut down before the application-level callbacks are made.  (End of advice to implementors.)"

What if the MPI implementation needs to add a callback sometime after MPI_Init based on what functions the user has called?

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-04-02 15:17:22 -0500**

---

Made a trivial modification in the proposal:

```
s/MPI_Init()\/MPI_Init_thread()/MPI_INIT \/ MPI_INIT_THREAD/
```

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-04-02 15:25:59 -0500**

---

Other than missing a changelog, this proposal looks good (read: we need a changelog entry ASAP before we can discuss it next week in Chicago!).

Rajeev -- I think the issue with enforced ordering was inspired by potentially using multiple different middleware packages that may not be able to coordinate their shutdown callbacks.  E.g., if package A uses package B which uses MPI, then you _need_ to have B's finalize callback called before A's.  A and B would be hard-pressed to coordinate this if they didn't know about each other.  And so on.

FWIW, Open MPI does _not_ invoke COMM_SELF callbacks in any particular order -- it'll actually be a little painful to convert to assign an order (because we use a hash table to store this stuff).  But it's all just bookkeeping -- it's certainly do-able.  So someone needs to do an open-source implementation.  

Bill -- did you get a chance to see if MPICH2 destroys COMM_SELF attributes in order?

**Comment by mpiforumbot** [comment]
**Originally by _rrtreu_ on 2009-04-03 08:00:02 -0500**

---

The MPI standard has always said attribute deletion callbacks occur in arbitrary order. In MPI-2 we added the requirement that the implementation do a pseudo free of MPI_COMM_SELF within MPI_FINALIZE to cause the execution of attribute deletion callbacks.

This was a sort of a hack but at least did not introduce explicit inconsistency. It served a real need. This ticket proposes adding explicitly inconsistent requirements when attribute deletion involves MPI_COMM_SELF within MPI_FINALIZE. Once again, it serves a real need but I am uncomfortable with the inconsistency.

It seems to me that defining a LIFO order for deleting attributes in general could be useful and can do no harm. I urge doing this in a consistent way by adding the mandate to all attribute deletions. 

I do not believe making LIFO a general mandate will harm any valid, existing MPI application and I do not believe it will be harder for implementators. 

Do not just add inconsistency to the standard.

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-04-03 14:09:24 -0500**

---

Added change log entry

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-04-03 14:14:54 -0500**

---

Replying to rsthakur:

> > Change  section 8.7.1 (p283, lines 43-44): from "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in an arbitrary order." to read "This will cause the delete callback function to be executed on all keys associated with MPI_COMM_SELF, in the reverse order they were set on MPI_COMM_SELF"
> 
> Can you not take care of the order by adding a single callback function for HDF and taking care of everything via extra_state?

Yes, that's what we do.  But, I need to make MPI calls during that callback and need to have the MPI library working correctly.

> > "Advice to implementors.  Since attributes can be added from any supported language, the MPI implementation needs to remember the creating language so the correct callback is made.  Implementations that make internal use of the attribute delete callback on MPI_COMM_SELF should register their internal callbacks before returning from MPI_Init/MPI_Init_Thread(), so that libraries or applications using the implementation will not have portions of the MPI implementation shut down before the application-level callbacks are made.  (End of advice to implementors.)"
> 
> What if the MPI implementation needs to add a callback sometime after MPI_Init based on what functions the user has called?

That should be disallowed by this implementation advise.  The middleware library (like HDF5) might require any aspect of the underlying MPI library and pieces of it should not be added/removed from "higher" in the callback stack by the MPI implementation itself.

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-04-03 14:16:20 -0500**

---

Tweaked change log

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-04-03 14:18:05 -0500**

---

Replying to rrtreu:

> The MPI standard has always said attribute deletion callbacks occur in arbitrary order. In MPI-2 we added the requirement that the implementation do a pseudo free of MPI_COMM_SELF within MPI_FINALIZE to cause the execution of attribute deletion callbacks.
> 
> This was a sort of a hack but at least did not introduce explicit inconsistency. It served a real need. This ticket proposes adding explicitly inconsistent requirements when attribute deletion involves MPI_COMM_SELF within MPI_FINALIZE. Once again, it serves a real need but I am uncomfortable with the inconsistency.
> 
> It seems to me that defining a LIFO order for deleting attributes in general could be useful and can do no harm. I urge doing this in a consistent way by adding the mandate to all attribute deletions. 
> 
> I do not believe making LIFO a general mandate will harm any valid, existing MPI application and I do not believe it will be harder for implementators. 
> 
> Do not just add inconsistency to the standard.

I'm happy to look up the changes required, but need some more experienced forum member's feedback on this idea.  Perhaps this should be a separate ticket?

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-04-06 13:59:26 -0500**

---

Ignore the first attachment - somehow I picked up the wrong file when browsing, and there seems to be no way to remove the attachment.  change-selfattr.txt is the one to look at.

This is a change against MPICH2 version 1.1b1 - I can post the full source if anyone is interested.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-04-06 14:27:59 -0500**

---

Bill -- there's a "delete attachment" button at the bottom of the attachment viewing page.  I took the liberty of deleting your 1st attachment.

**Comment by mpiforumbot** [comment]
**Originally by _bronis_ on 2009-04-06 14:45:23 -0500**

---

Please change "order they" to "order that they"

Please change "Implementations that make internal use of the attribute delete callback on MPI_COMM_SELF should register their internal callbacks before returning from MPI_INIT / MPI_INIT_THREAD, so that libraries or applications using the implementation will not have portions of the MPI implementation shut down before the application-level callbacks are made." to "Implementations that use the attribute delete callback on MPI_COMM_SELF internally should register their internal callbacks before returning from MPI_INIT / MPI_INIT_THREAD, so that libraries or applications will not have portions of the MPI implementation shut down before the application-level callbacks are made." (tweak wording in first clause and delete "using the implementation", which seems unnecessary).

Both of these changes are very minor and should not reset the review count.

Overall, this looks good generally and this can count as my review.

**Comment by mpiforumbot** [comment]
**Originally by _koziol_ on 2009-04-06 14:53:32 -0500**

---

Incorporate Bronis' grammar suggestions.

Mark as fully reviewed now.

Mark as having implementation (from Bill's patch).

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-04-06 14:54:25 -0500**

---

**Attachment added:** [`change-selfattr.diff`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/71/change-selfattr.diff) (4.3 KiB)
Re-attach same old attachment, but with a .diff extension so that trac shows it nicely

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2009-04-06 14:55:02 -0500**

---

I took the further liberty of deleting Bill's 2nd attachment and re-attaching it with a .diff extension so that Trac will show it nicely.

**Comment by mpiforumbot** [comment]
**Originally by _asupalov_ on 2009-04-07 12:48:55 -0500**

---

Putting into writing the concern I expressed verbally during the meeting:

From the point of view of a commercial MPI vendor, changing the order in which the callbacks are fired may lead to malfunctioning certain applications that incorrectly relied on a certain order used by their favourite MPI implementation.

The most lively example of what may happen here is the MPI_Send to self without a pending MPI_Irecv. Experience shows that what typically happens is the following:
1. The MPI implementation that "breaks" the application is blamed for this. This take weeks to clarify.
2. When 1. is adequately addressed by cross-MPI testing, the MPI standard is blamed for this. This takes months to get into ISV's mind.
3. When 2. is abundantly clarified, ISV may accept that their application is broken and go fix it. This takes years to reach the marketplace.

Most often, the discussion withers well before 3, in which case the MPI implementation at hand is subjected to increased competitive pressure created by other, non-complying MPI implementations that "do the right thing" - i.e., preserve compatibility with the older version of the MPI standard by not changing the callback firing sequence.

As a risk mitigation procedure, wise MPI implementors always include backward compatibility mode into their product. This should probably be mentioned in the "effect" part of the ticket.

Without due diligience here, we may increase the barrier to MPI-2.2 adoption by both commercial MPI vendors and commercial ISVs.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-07-27 16:11:22 -0500**

---

Removed the "probably should"

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-08-03 13:08:49 -0500**

---

**Attachment added:** [`change-log_ticket71_item22.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/71/change-log_ticket71_item22.pdf) (396.5 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _bosilca_ on 2009-08-14 17:28:15 -0500**

---

**Attachment added:** [`inquiry-ticket71.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/71/inquiry-ticket71.pdf) (106.1 KiB)
Changes for ticket 71

**Comment by mpiforumbot** [comment]
**Originally by _bosilca_ on 2009-08-14 17:29:20 -0500**

---

Revision r366 implement the fix required by this ticket.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2009-08-29 14:32:40 -0500**

---

PDF review: okay.

**Comment by mpiforumbot** [comment]
**Originally by _gropp_ on 2009-09-03 00:48:21 -0500**

---

Master copy reviewed and ok.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2010-09-18 04:58:06 -0500**

---

This ticket is (long-since) complete; marking it closed.
